### PR TITLE
 Fixed #28837 -- Fixed test client crash if an exception with more than one arg is raised.

### PR DESCRIPTION
**Pull request by ticosax**
In case the exception instance accept more than one positional parameter
test client was failing to re-raise it.

Fix borrowed from `six.reraise` implementation

### CODE REVIEW

**Comment by ticosax** [comment]
It is a regression noticed while running our test suite against 2.0rc1. If accepted it needs to be backported.

ref commit https://github.com/django/django/commit/6e55e1d88a5c4453e25f0caf7ffb68973de5c0ba

**Comment by charettes** [comment]
@ticosax that looks like a legitimate issue. Could you please file a Trac ticket and mark it as a release blocker against 2.0.

**Comment by ticosax** [comment]
thank you @charettes. I've done it.

**Comment by charettes** [comment]
@ticosax thanks, left some comments on the ticket regarding other possible instances of the issue introduced by 6e55e1d88a5c4453e25f0caf7ffb68973de5c0ba.

File: django/utils/sys.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+def reraise(exc_type, exc_value, tb=None):
+    """re raise excpetions captured by sys.exec_info()"""
+    try:
+        if exc_value is None:
+            exc_value = exc_type()
+        if exc_value.__traceback__ is not tb:
+            raise exc_value.with_traceback(tb)
+        raise exc_value
+    finally:
```
Comment (timgraham) [inline django/utils/sys.py]: I don't see the purpose of the try/finally (this is not in `six.reraise()` and no tests fail if it's removed. Further, the if statements are never evaluating to True.

File: django/utils/sys.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+def reraise(exc_type, exc_value, tb=None):
+    """re raise excpetions captured by sys.exec_info()"""
```
Comment (timgraham) [inline django/utils/sys.py]: The docstring doesn't seem very descriptive. Maybe you could adapt from [six's docs](https://pythonhosted.org/six/#six.reraise)? Probably unit tests for `reraise()` should be added.

File: django/utils/sys.py (hunk 1 above)
Comment (ticosax) [reply django/utils/sys.py]: Looking up for six project, I ended on [bitbucket](https://bitbucket.org/gutworth/six/src/92e1c746e1abfbdb94705b821bc93766083efc54/six.py?at=default&fileviewer=file-view-default) former repository,
It explain the discrepancy, I will fix it asap.
Thank you.

File: django/utils/sys.py (hunk 2 above)
Comment (ticosax) [reply django/utils/sys.py]: will do Now I found upstream project

Comment (ticosax) [reply django/utils/sys.py]: https://github.com/benjaminp/six/blob/master/six.py#L687
There is no docstring in upstream six. Please, let me know what you want to read.

File: django/utils/sys.py (hunk 1 above)
Comment (ticosax) [reply django/utils/sys.py]: looks like upstream six repo is still using the same implementation.
I wouldn't change it. would you ?

Comment (ticosax) [reply django/utils/sys.py]: > no tests fail if it's removed.

it fails for me.
```
Traceback (most recent call last):
  File "tests/runtests.py", line 16, in <module>
    from django.test import TestCase, TransactionTestCase
  File "/home/nicolas/src/django/django/test/__init__.py", line 3, in <module>
    from django.test.client import Client, RequestFactory
  File "/home/nicolas/src/django/django/test/client.py", line 28, in <module>
    from django.utils.sys import reraise
ImportError: cannot import name 'reraise'
```
I will add a test for `reraise()` explicitly though.

Comment (timgraham) [reply django/utils/sys.py]: What I meant is that if you change the implementation to: 
```
def reraise(exc_type, exc_value, tb=None):
    raise exc_value
```
the new `test_client` test still passes. It also passes with `raise exc_info[1].with_traceback(exc_info[2])` as Claude suggested in the ticket. I think there should be additional assertions in the `test_client` test to demonstrate which implementation is correct.

Comment (ticosax) [reply django/utils/sys.py]: >  It also passes with raise exc_info[1].with_traceback(exc_info[2]) 

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.5/unittest/case.py", line 600, in run
    testMethod()
  File "/home/nicolas/src/django/tests/test_client/tests.py", line 730, in test_response_raise_exception
    self.client.get('/error/')
  File "/home/nicolas/src/django/django/test/client.py", line 518, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File "/home/nicolas/src/django/django/test/client.py", line 333, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/nicolas/src/django/django/test/client.py", line 405, in generic
    return self.request(**r)
  File "/home/nicolas/src/django/django/test/client.py", line 486, in request
    reraise(*exc_info)
  File "/home/nicolas/src/django/django/utils/sys.py", line 3, in reraise
    raise exc_type(exc_value).with_traceback(tb)
TypeError: __init__() missing 1 required positional argument: 'also_required'
```
Also with the test I backported from six (last commit), the `reraise` function is fully covered

Comment (timgraham) [reply django/utils/sys.py]: Yes, but the if statements don't evaluate to True in the test_client test so it's unclear whether they are  actually needed.

Comment (ticosax) [reply django/utils/sys.py]: Would you take the risk of using something else than `six` is doing ?
What if other corner cases comes up later for someone else.
This test proves that `raise exc_info[0](exc_info[1]).with_traceback(exc_info[2])` is not bullet proof. I would rather rely on a more trusted implementation from `six`, even if we don't know how to reproduce all the corner cases within the test_client test suite.
We know the actual implementation is fragile, since the fix is not trivial, I 'm incline to blindly trust `six.reraise` implementation since it is what was used before 2.0.

Comment (timgraham) [reply django/utils/sys.py]: I'd like to understand the code (and simplify, if possible) rather than include code that may not be necessary. I'll take a look if you don't want to spend more time on the issue.

Comment (ticosax) [reply django/utils/sys.py]: I don't want to make your work more difficult. please proceed.

File: django/test/client.py [hunk 3]
```diff
@@ -482,7 +483,7 @@ def request(self, **request):
             if self.exc_info:
                 exc_info = self.exc_info
                 self.exc_info = None
-                raise exc_info[0](exc_info[1]).with_traceback(exc_info[2])
```
Comment (timgraham) [inline django/test/client.py]: @claudep, as far as I see, `raise exc_info[1]` works here. I'm not if `with_traceback(exc_info[2])` ever adds anything. Do you have a good understand of its purpose? 

Here's a sample traceback which is the same output as older versions of Django:
```
======================================================================
ERROR: test_detail_view_with_a_past_question (polls.tests.QuestionIndexDetailTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/mysite/polls/tests.py", line 128, in test_detail_view_with_a_past_question
    response = self.client.get(reverse('polls:detail', args=(past_question.id,)))
  File "/home/tim/code/django/django/test/client.py", line 519, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File "/home/tim/code/django/django/test/client.py", line 333, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/tim/code/django/django/test/client.py", line 405, in generic
    return self.request(**r)
  File "/home/tim/code/django/django/test/client.py", line 486, in request
    raise exc_info[1]
  File "/home/tim/code/django/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/home/tim/code/django/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/tim/code/django/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/tim/code/django/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/tim/code/django/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/tim/code/django/django/views/generic/detail.py", line 105, in get
    self.object = self.get_object()
  File "/home/tim/code/django/django/views/generic/detail.py", line 30, in get_object
    queryset = self.get_queryset()
  File "/home/tim/code/mysite/polls/views.py", line 42, in get_queryset
    raise CustomExceptionWith2PositionalParam('required', 'also_required')
polls.views.CustomExceptionWith2PositionalParam: ('required', 'also_required')
```
For the test that's added in this PR, the if statements in `six.reraise()` don't evaluate to True, so it does `raise value` (which is the same thing as `raise exc_info[1]`).
```
def reraise(tp, value, tb=None):
    if value is None:
        value = tp()
    if value.__traceback__ is not tb:
        raise value.with_traceback(tb)
    raise value
```

Comment (timgraham) [reply django/test/client.py]: Looking at [the C source code](https://github.com/python/cpython/blob/4d193bcc2560b824389e4d98d9d8b9b34e33dbaf/Objects/exceptions.c#L161-L173), `with_traceback(tb)` does `self.__traceback_ = tb`. The `with_traceback(...)` part of `exc_info[1].with_traceback(exc_info[2])` looks useless because  adding this assertion before that line always passes: `assert exc_info[1].__traceback__ is exc_info[2]`. I guess the `with_traceback()` addition was meant to emulate the `if value.__traceback__ is not tb: raise value.with_traceback(tb)` from `six.reraise()` but I don't see how that code could run.

Comment (claudep) [reply django/test/client.py]: Thanks for investigating, so simply raising `exc_info[1]` seems the way to go, unless some test can demonstrate the opposite.

File: django/db/models/query.py [hunk 4]
```diff
@@ -521,13 +520,12 @@ def _create_object_from_params(self, lookup, params):
...
             return obj, True
-        except IntegrityError:
-            exc_info = sys.exc_info()
+        except IntegrityError as e:
             try:
                 return self.get(**lookup), False
             except self.model.DoesNotExist:
                 pass
-            raise exc_info[0](exc_info[1]).with_traceback(exc_info[2])
+            raise e
```
Comment (charettes) [inline django/db/models/query.py]: I don't think the explicit `raise e` is required on Python 3.5+.
```python
(tmp-70aba34da671918e) ➜  tmp-70aba34da671918e ipython
Python 3.5.2 (default, Sep 17 2016, 14:04:19)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: def func():
   ...:     try:
   ...:         raise KeyError('foo')
   ...:     except KeyError:
   ...:         try:
   ...:             raise Exception('test')
   ...:         except Exception:
   ...:             pass
   ...:         raise
   ...:

In [2]: func()
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-bd1982955a12> in <module>()
----> 1 func()

<ipython-input-1-e80913097bf5> in func()
      1 def func():
      2     try:
----> 3         raise KeyError('foo')
      4     except KeyError:
      5         try:

KeyError: 'foo'
```

Comment (timgraham) [reply django/db/models/query.py]: Yes, I remember `raise` vs. `raise e` being a point of confusion in the past. If it's just `raise`, then lines 3-4 below are omitted in the traceback:
```
File "/home/tim/code/django/django/db/models/query.py", line 489, in get_or_create
  return self._create_object_from_params(lookup, params)
File "/home/tim/code/django/django/db/models/query.py", line 528, in _create_object_from_params
  raise e
File "/home/tim/code/django/django/db/models/query.py", line 521, in _create_object_from_params
  obj = self.create(**params)
```
It could possibly be unclear without those lines, but I'm not sure if it matters.

Comment (charettes) [reply django/db/models/query.py]: Makes sense, let's stick with `raise e`.
